Add an Azure Key Vault KMS implementation

robobario opened the pull request:
### Type of change

- Enhancement / new feature

### Description

We support pluggable KMS implementations for the Record Encryption Filter. This adds support for Azure Key Vault and it's related variant Managed HSM.

## Features
1. Support `RSA` and `RSA-HSM` key types, wrapping using `RSA-OAEP-256` but emit a warning that it is not quantum-resistant.
2. Support `oct` and `oct-HSM` key type with `AES-GCM` wrapping algorithm (will only work with Managed HSM)
3. Support only client credentials oauth flow with Entra using clientId
   + clientSecret. This supports workloads running anywhere.
4. Support TLS customization of the authentication client and key vault client.
5. DEK bytes are generated proxy-side with a SecureRandom
6. Edek stores the keyName, keyVersion, edek.
7. User will supply tenantId for authentication. It is possible to obtain this by making an HTTP request to Key Vault with no Authorization header. But it should be an easy thing for users to supply.
8. An alias is 1:1 with a key name in Key Vault.
9. When we resolve the alias we fail fast if:
   - the keytype is not supported
   - the key is not enabled
   - the key operations do not contain both wrapKey and unwrapKey

## Auth:
With the oauth client credentials flow we obtain a Bearer token along with a remaining lifetime in seconds (default 1 hour). We share the token across all filters for a single Filter definition. When it gets near to expiry (1 minute away), then we trigger an asynchronous refresh. The established token continues to be used while the refresh is in progress.

## Example configuration:
```yaml
kms: AzureKeyVault
kmsConfig:
  keyVaultBaseUrl: https://kv-my-vault-name.vault.azure.net
  tls:
    ... client trust for key vault
  entraIdentity:
     oauthEndpointUrl: https://login.microsoftonline.com // optional defaults to this value
     clientId:
       passwordFile: /path/to/id
     clientSecret:
       passwordFile: /path/to/secret
     tenantId: "abds-1232dsaa"
     scope: https://vault.azure.net/.default // optional, defaults to this value
     tls:
       ... client trust configuration for oauth
```
### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] PR raised from a fork of this repository and made from a branch rather than main. 
- [ ] Write tests
- [ ] Update documentation
- [ ] Make sure all unit/integration tests pass
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] If applicable to the change, [trigger the system test suite](../blob/main/DEV_GUIDE.md#jenkins-pipeline-for-system-tests).  Make sure tests pass.
- [ ] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [ ] Ensure the PR references relevant issue(s) so they are closed on merging.
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

robobario commented:
Hi @k-wall I ran out of time to get the integration testing support done. But it's already a bit of a monster to get things going end-to-end with authentication etc, maybe better if we add it after.

robobario commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/keyvault/SupportedKeyType.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.Optional;
+
+public enum SupportedKeyType {
```
this is a way of classifying some key types as there are `RSA` and `RSA-HSM` variants that aren't really different from our point of view, they will use the same wrap algorithm.

But I'm wondering if it would be better to keep it closer to what Azure exposes. So have a `KeyType` enum and an algorithm enum that matches Azure and operate on those.

robobario commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/AzureKeyVaultEdekSerde.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Override
+    public int sizeOf(AzureKeyVaultEdek edek) {
+        return edek.keyVersion128bit().map(e -> versionAsHexBytesSize(edek)).orElse(versionAsStringSize(edek));
```
this is something I've been wondering about.

Azure tells us the key id, which looks like `https://myvault.vault.azure.net/keys/CreateSoftKeyTest/78deebed173b48e48f55abf87ed4cf71` ([documented](https://learn.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets-certificates#object-identifiers) as being case-insensitive). This `id` happens to be usable when you plug it in as the prefix to wrap/unwrap URLs, the Azure SDKs operate like this.

We are only storing the `CreateSoftKeyTest` and `78deebed173b48e48f55abf87ed4cf71` portions to try and reduce the size of the overhead per record.

The version part is only defined in their docs as a 32-length string. So we must be pessimistic and tolerate the worst case of 32 unicode characters.

However, all examples are 32 lowercase hex characters. So I've made it optimistically test if it's lowercase hex and we can decode 16 bytes of data, then we store that as a v0 EDEK. Else we store it as a v1 EDEK. We require lowercase for fidelity reasons, I want to only make this optimization if we can guarantee that we can decode it back to the same String that came out of the Azure API.

This enables us save around 17 bytes per record as we don't have to store the length of the version bytes and can use fewer bytes for the data.

robobario commented:
mm the lowkey/azure client libraries is causing a bit of mischief, they pulls in a lot of netty libs including `netty-tcnative-boringssl-static` which is changing the proxy TLS behaviour in `kroxylicious-integration-tests`.

It looks like we can get away with excluding `com.azure:azure-core-http-netty` from being pulled in which removes all the netty dependencies. Azure supports pluggable http clients and the 1st part implementation is netty-based. Lwkey also provides an apache httpclient backed implementation of the azure HttpClient with specializations like being able to trust self-signed certs and manipulate URI authorities, so we can exclude the netty implementation and only use the apache implementation.

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms-test-support/pom.xml:
```diff
@@ -0,0 +1,127 @@
+        <dependency>
+            <groupId>com.github.nagyesta.lowkey-vault</groupId>
+            <artifactId>lowkey-vault-testcontainers</artifactId>
+            <version>4.0.0</version><exclusions>
```
extract a property for this?

robobario replied:
have shuffled the new deps up into the parent and used properties for the versions

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms-test-support/src/main/java/io/kroxylicious/kms/provider/azure/kms/AzureKeyVaultKmsTestKmsFacade.java:
```diff
@@ -0,0 +1,233 @@
+        this.kms = startVault();
+        try {
+            this.entraMock = HttpServer.create();
+            entraMock.createContext("/", exchange -> {
```
Can we use WireMockServer here instead or rolling our own?  We've used WireMockServer extensively already and I think it would be better not to diverge.

robobario replied:
it's all wiremock now

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms-test-support/src/main/java/io/kroxylicious/kms/provider/azure/kms/AzureKeyVaultKmsTestKmsFacade.java:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    public LowkeyVaultContainer startVault() {
+        final DockerImageName imageName = DockerImageName.parse("nagyesta/lowkey-vault:4.0.0");
```
Is this available in the google mirror?
We should add a renovate rule for this container image too.

robobario replied:
looks like it is:

```
podman pull mirror.gcr.io/nagyesta/lowkey-vault:4.0.0
Trying to pull mirror.gcr.io/nagyesta/lowkey-vault:4.0.0...
..
b2d9fda7634f9b9b2e6437739a139a50f555e85738d31253d948b5ef45d0f309
```

robobario replied:
makes me wonder if we need a kroxylicious-testcontainers module to hold extensions like the Image name substitutor. The integration suite is already pulling from the mirror, but the other modules aren't. I've made them directly reference the mirror for now.

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/pom.xml:
```diff
@@ -0,0 +1,146 @@
+    </parent>
+
+    <artifactId>kroxylicious-kms-provider-azure-key-vault-kms</artifactId>
+    <packaging>jar</packaging>
```
I think jar is the default, no?

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/AzureKeyVaultKms.java:
```diff
@@ -0,0 +1,193 @@
+
+    @NonNull
+    private static SupportedKeyType validateKeyAndExtractType(String alias, GetKeyResponse response, JsonWebKey key) {
+        if (REQUIRED_OPERATIONS.stream().anyMatch(p -> !key.keyOperations().contains(p))) {
```
nice!

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/test/java/io/kroxylicious/kms/provider/azure/auth/OauthClientCredentialsTokenServiceTest.java:
```diff
@@ -0,0 +1,127 @@
+        Clock clock = Clock.fixed(fixedInstant, ZoneId.of("UTC"));
+        int expirySeconds = 3200;
+        Responder resp = exchange -> {
+            var response = Map.of("token_type", "Bearer", "expires_in", expirySeconds, "access_token", TOKEN);
```
When writing unit tests like these, I try to use the responses copied straight from the documentation of the services, with as few changes as possible, so we are testing against responses from an authoritative known good source.
So in this case, I'd copy the response from:

https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#successful-response-1

I think if you don't do this there's more risk that the code in your own misunderstandings of the API into the responses you hand-craft.

See io.kroxylicious.kms.provider.aws.kms.credentials.Ec2MetadataCredentialsProviderTest#KNOWN_GOOD_SECURITY_CREDENTIAL_RESPONSE

robobario replied:
sounds good, slight hassle that they've used ellipsis to reduce the size of the tokens and made their examples slightly less realistic.

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/test/java/io/kroxylicious/kms/provider/azure/keyvault/KeyVaultClientIT.java:
```diff
@@ -0,0 +1,142 @@
+            when(bearerTokenService.getBearerToken()).thenReturn(CompletableFuture.completedFuture(token));
+            EntraIdentityConfig unused = new EntraIdentityConfig(null, "tenant", new InlinePassword("abc"), new InlinePassword("def"), null, null);
+            KeyVaultClient keyVaultClient = new KeyVaultClient(bearerTokenService,
+                    new AzureKeyVaultConfig(unused, lowkeyVaultContainer.getVaultBaseUrl(VAULT_NAME), INSECURE_TLS));
```
`lowkeyVaultContainer.getVaultBaseUrl(VAULT_NAME)` fails for me because it generates a url with a hostname that doesn't resolve.

<img width="2454" height="1381" alt="image" src="https://github.com/user-attachments/assets/56b19273-b1a0-4da5-abac-acc453862ce8" />

k-wall commented:
Thanks @robobario, I think this PR is going in the right direction.  I've left a couple of comments.  I'd like to take a deeper look at the authentication stuff.  I would like us to switch to WireMock so we have the uniformity of approach.

robobario commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/AzureKeyVaultKms.java:
```diff
@@ -0,0 +1,193 @@
+                LOG.debug("resolved alias {} to {}", alias, response);
+                JsonWebKey key = response.key();
+                SupportedKeyType keyType = validateKeyAndExtractType(alias, response, key);
+                if (!keyType.isQuantumResistant()) {
```
while in the `SupportedKeyType` area, I think I will also change it so that there is an enum element for each of `RSA`, `RSA-HSM`, `oct` and `oct-HSM` matching the API, rather than having our own SupportedKeyType, I don't think there's a need to group them together into RSA-like and AES-like, better to stay close to the API.

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/auth/BearerToken.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public String toString() {
+        return "BearerToken{" +
+                "token='REDACTED'" +
```
```suggestion
                "token='*****'" +
```

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/config/AzureKeyVaultConfig.java:
```diff
@@ -0,0 +1,44 @@
+
+@JsonPropertyOrder({ "entraIdentity", "keyVaultBaseUrl", "tls" })
+public record AzureKeyVaultConfig(@JsonProperty(required = true) EntraIdentityConfig entraIdentity,
+                                  @JsonProperty(required = true) String keyVaultBaseUrl,
```
Jackson knows about URI, and we've used it on Configuration objects pretty extensively already.

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/config/auth/EntraIdentityConfig.java:
```diff
@@ -0,0 +1,54 @@
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+@JsonPropertyOrder({ "oauthEndpoint", "tenantId", "clientId", "clientSecret", "scope", "tls" })
+public record EntraIdentityConfig(@JsonInclude(JsonInclude.Include.NON_NULL) @Nullable String oauthEndpoint,
```
ditto - make oauthEndpoint a URI

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/config/auth/EntraIdentityConfig.java:
```diff
@@ -0,0 +1,54 @@
+        Objects.requireNonNull(clientId, "clientId cannot be null");
+        Objects.requireNonNull(clientSecret, "clientSecret cannot be null");
+        if (oauthEndpoint != null && !oauthEndpoint.startsWith("https://")) {
+            LOG.warn("oauthEndpoint {} does not begin with https://, production installations should use a secure endpoint", oauthEndpoint);
```
not something for now, but I wonder if we could make the kroxylicious walk the config model and bleat about any non-HTTPs URI it finds (we'd need an exception for AWS EC2)

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/keyvault/KeyVaultClient.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    private HttpRequest getKeyRequest(String keyName, BearerToken bearerToken) {
+        String getKey = config.keyVaultBaseUrl() + "/keys/" + keyName + "?api-version=7.4";
```
we should pull out the api-version as a constant.

k-wall commented on pom.xml:
```diff
@@ -360,13 +365,46 @@
                 <artifactId>api</artifactId>
                 <version>${strimzi.version}</version>
             </dependency>
+            <dependency>
```
We want to avoid these libraries becoming prod dependencies by mistake.  I think it this case, I'd manage these at the child pom level, or make their scope test here, to make the intent clear.

robobario commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/auth/OauthClientCredentialsTokenService.java:
```diff
@@ -0,0 +1,133 @@
+        tlsConfigurator.apply(builder);
+        httpClient = builder
+                .connectTimeout(Duration.of(10, SECONDS))
+                .followRedirects(HttpClient.Redirect.NEVER)
```
My thought was that we should be very selective about where we're posting the credentials. But reading into it a bit further I think this is also aligned better with the oauth2 spec:

[RFC 6479](https://datatracker.ietf.org/doc/html/rfc6749#section-5) lays out only a 200 success and a 400 or 401 error response. So I think a failure on any other code is the right move to keep within the defined flows.

robobario commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/keyvault/GetKeyResponse.java:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * @return if this key type is supported, return it's key type, else empty
+     */
+    public Optional<SupportedKeyType> supportedKeyType() {
```
yep, better for this to just model the API response. I have moved the supported-key-type logic up to a higher level

k-wall commented:
If you make a mistake configuring Entra, you don't get much clue from the logs, even at DEBUG.

```
Caused by: io.kroxylicious.kms.provider.azure.UnexpectedHttpStatusCodeException: response has an unexpected status code: 400
	at io.kroxylicious.kms.provider.azure.auth.OauthClientCredentialsTokenService.getAccessTokenResponse(OauthClientCredentialsTokenService.java:118)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
```

if you have UnexpectedHttpStatusCodeException include the response body, you get some useful stuff:

```
Caused by: java.util.concurrent.CompletionException: io.kroxylicious.kms.provider.azure.UnexpectedHttpStatusCodeException: response has an unexpected status code: 400{"error":"unauthorized_client","error_description":"AADSTS700016: Application with identifier 'a130cb56-8604-40ea-909a-d6eb5505642f' was not found in the directory 'Default Directory'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant. Trace ID: 4fef1deb-1872-422a-8fc1-a2ac6ff72e00 Correlation ID: 8be835bd-be0e-467e-8351-49bd3b8be62e Timestamp: 2025-09-24 14:40:17Z","error_codes":[700016],"timestamp":"2025-09-24 14:40:17Z","trace_id":"4fef1deb-1872-422a-8fc1-a2ac6ff72e00","correlation_id":"8be835bd-be0e-467e-8351-49bd3b8be62e","error_uri":"https://login.microsoftonline.com/error?code=700016"}
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649)

```

k-wall commented:
Wanted to share a couple of things  that I needed to do to test against real Azure:

<img width="2329" height="594" alt="image" src="https://github.com/user-attachments/assets/801f7d6f-b5e0-4ceb-a7f5-31266778b0fa" />

k-wall commented:
We should probably have DEV_GUIDEs for the KMS showing the developer the minimum steps they need to actually test the KMS implementation against the real service.

robobario commented:
I've refactored it so that we store the `vaultName` and `keyType` in the serialized EDEK and use these values to execute the unwrap operation without having to get the key details again. So the decrypt side is simpler, with just the one API call.

With this change the host name became more meaningful, ie `keyVaultBaseUrl: https://my-kv.vault.azure.net` started to encode both a key vault name for encryption (`my-kv`) and a sort of base vault host (`vault.azure.net`) which we want to use at unwrap time, prepending the vaultName from the serialized EDEK to the vault host.

So, the existing `keyVaultBaseUrl` was fighting against this new design. I've split it up into some new properties:

* `keyVaultName` required - defines the key vault name to use for encryption.
* `keyVaultHost` required - defines the base host which vault names are prepended to. for example, `vault.azure.net`.
* `keyVaultScheme` optional - default `https`, enables tests to override it to http
* `keyVaultPort` optional - default `null`, enables tests to configure a port. If omitted we will not specify a port when constructing requests to vault.
* `omitVaultNameFromHost` optional, default false. Enables integration tests to function if the system can only resolve `localhost`. 

As you can see there is a downside, since we use `lowkey` in a mode where we address it with `https://localhost:45678/` as a default vault endpoint, this doesn't conform to our new expectation of how things work. Lowkey can operate with urls like  `https://{vault-name}.localhost:45678/` but we found that not all systems can resolve these hosts. So we need some specialization to omit the vault-name when constructing requests. It means that we cannot easily test that decryption works across vault names. We would need a test environment that can use a realistic looking vault addressing scheme.

k-wall commented on kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/main/java/io/kroxylicious/kms/provider/azure/config/AzureKeyVaultConfig.java:
```diff
@@ -0,0 +1,105 @@
+/**
+ * @param entraIdentity required credentials for authenticating with Entra
+ * @param keyVaultName required name of the key vault to use for encryption, e.g. my-key-vault
+ * @param keyVaultHost required host of key vault (without key vault name) e.g. vault.azure.net
```
I see your explanation now.
